### ISSUE TITLE
Script URL step with HTTP based uri fails on multiple nodes, successfull on a single node

### ISSUE THREAD

**Issue by GolubevV** [rundeck/rundeck#1055]
Configuration:
- rundeck
  - base	/var/lib/rundeck
  - build	2.3.2-1
- os
  - arch	amd64
  - name	Linux
  - version	2.6.32-431.29.2.el6.centos.plus.x86_64
- jvm
  - implementationVersion	23.25-b01
  - name	OpenJDK 64-Bit Server VM
  - vendor	Sun Microsystems Inc.
  - version	1.6.0_33

Conditions:
* use Script URL step as part of the job definition;
* specify the URL of the script to run in the HTTP-based URI ( http://host/path_to_script/script.sh);

If job nodes filter expand to multiple nodes, the first node gets the requested script downloaded to its temp folder, while other node(s) have the following error stacktrace:

^2014-12-15T14:42:39Z||VERBOSE|{node=localhost|step=1|stepctx=1|user=_some_user_}|Failed dispatching to node _some_node_:
````java
com.dtolabs.rundeck.core.execution.workflow.steps.node.NodeStepException: [jsch-scp] Failed copying the file: java.io.FileNotFoundException: /var/lib/rundeck/var/cache/ScriptURLNodeStepExecutor/87d6cff64dad7d6fe11f182d8b439a048f2a2f47.temp (No such file or directory)
        at com.dtolabs.rundeck.core.execution.workflow.steps.node.impl.ScriptFileNodeStepExecutor.executeScriptFile(ScriptFileNodeStepExecutor.java:138)
        at com.dtolabs.rundeck.core.execution.workflow.steps.node.impl.ScriptURLNodeStepExecutor.executeNodeStep(ScriptURLNodeStepExecutor.java:111)
        at com.dtolabs.rundeck.core.execution.ExecutionServiceImpl.executeNodeStep(ExecutionServiceImpl.java:149)
        at com.dtolabs.rundeck.core.execution.dispatch.SequentialNodeDispatcher.dispatch(SequentialNodeDispatcher.java:116)
        at com.dtolabs.rundeck.core.execution.dispatch.SequentialNodeDispatcher.dispatch(SequentialNodeDispatcher.java:58)
        at com.dtolabs.rundeck.core.execution.ExecutionServiceImpl.dispatchToNodes(ExecutionServiceImpl.java:177)
        at com.dtolabs.rundeck.core.execution.workflow.steps.NodeDispatchStepExecutor.executeWorkflowStep(NodeDispatchStepExecutor.java:66)
        at com.dtolabs.rundeck.core.execution.ExecutionServiceImpl.executeStep(ExecutionServiceImpl.java:116)
        at com.dtolabs.rundeck.core.execution.workflow.BaseWorkflowStrategy.executeWFItem(BaseWorkflowStrategy.java:165)
        at com.dtolabs.rundeck.core.execution.workflow.BaseWorkflowStrategy.executeWorkflowItemsForNodeSet(BaseWorkflowStrategy.java:225)
        at com.dtolabs.rundeck.core.execution.workflow.BaseWorkflowStrategy.executeWorkflowItemsForNodeSet(BaseWorkflowStrategy.java:192)
        at com.dtolabs.rundeck.core.execution.workflow.StepFirstWorkflowStrategy.executeWorkflowImpl(StepFirstWorkflowStrategy.java:79)
        at com.dtolabs.rundeck.core.execution.workflow.BaseWorkflowStrategy.executeWorkflow(BaseWorkflowStrategy.java:123)
        at com.dtolabs.rundeck.core.execution.workflow.NodeFirstWorkflowStrategy$DispatchedWorkflow.dispatch(NodeFirstWorkflowStrategy.java:366)
        at com.dtolabs.rundeck.core.execution.dispatch.ParallelNodeDispatcher$1.call(ParallelNodeDispatcher.java:179)
        at com.dtolabs.rundeck.core.cli.CallableWrapperTask.execute(CallableWrapperTask.java:52)
        at sun.reflect.GeneratedMethodAccessor1604.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:622)
        at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
        at org.apache.tools.ant.Task.perform(Task.java:348)
        at org.apache.tools.ant.taskdefs.Parallel$TaskRunnable.run(Parallel.java:428)
        at java.lang.Thread.run(Thread.java:701)
Caused by: com.dtolabs.rundeck.core.execution.service.FileCopierException: [jsch-scp] Failed copying the file: java.io.FileNotFoundException: /var/lib/rundeck/var/cache/ScriptURLNodeStepExecutor/87d6cff64dad7d6fe11f182d8b439a048f2a2f47.temp (No such file or directory)
        at com.dtolabs.rundeck.core.execution.impl.jsch.JschScpFileCopier.copyFile(JschScpFileCopier.java:179)
        at com.dtolabs.rundeck.core.execution.impl.jsch.JschScpFileCopier.copyFile(JschScpFileCopier.java:209)
        at com.dtolabs.rundeck.core.execution.ExecutionServiceImpl.fileCopyFile(ExecutionServiceImpl.java:261)
        at com.dtolabs.rundeck.core.execution.workflow.steps.node.impl.ScriptFileNodeStepExecutor.executeScriptFile(ScriptFileNodeStepExecutor.java:131)
        ... 22 more
Caused by: java.io.FileNotFoundException: /var/lib/rundeck/var/cache/ScriptURLNodeStepExecutor/87d6cff64dad7d6fe11f182d8b439a048f2a2f47.temp (No such file or directory)
        at org.apache.tools.ant.taskdefs.optional.ssh.Scp.execute(Scp.java:215)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:622)
        at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:106)
        at org.apache.tools.ant.Task.perform(Task.java:348)
        at org.apache.tools.ant.taskdefs.Sequential.execute(Sequential.java:62)
        at com.dtolabs.rundeck.core.execution.impl.jsch.JschScpFileCopier.copyFile(JschScpFileCopier.java:169)
        ... 25 more
````

If looking at the temp folder of the target nodes, the file is present on one node, and is empty on others:

**Node1** /tmp listing:
-rwxr-xr-x 1 rundeck       rundeck       2477 Dec 15 15:41 1418654559187-_node1_-87d6cff64dad7d6fe11f182d8b439a048f2a2f47.temp.sh

**Node2** /tmp listing:
-rw-r--r-- 1 rundeck rundeck    0 Dec 15 15:42 1418654559188-_node2_-87d6cff64dad7d6fe11f182d8b439a048f2a2f47.temp.sh

If node filter of the same job definitions expands just to single node, the step works correctly.
So it looks like the Rundeck server downloads script from the configured URI and moves it  (instead of copying) to the first node in the list, resulting in errors when trying to copy to all other nodes.

**Comment by gschueler** [comment]
can you verify this issue is solved in 2.4.0 release?

**Comment by GolubevV** [comment]
Sorry not to respond earlier, the issue is fixed indeed in 2.4.0 release
